src/main/java/com/example/Ecommerce/service/CartItemQueueService.java

```diff
@@ -0,0 +1,80 @@
+package com.example.Ecommerce.service;
+
+import com.example.Ecommerce.entity.*;
+import com.example.Ecommerce.exception.ResourceNotFoundException;
+import com.example.Ecommerce.model.QueueItem;
+import com.example.Ecommerce.repository.CartItemRepo;
+import com.example.Ecommerce.repository.ProductRepo;
+import com.example.Ecommerce.util.NotificationClient;
+import lombok.RequiredArgsConstructor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemQueueService {
+
+    private final ProductRepo productRepo;
+    private final CartItemRepo cartItemRepo;
+    private final NotificationClient notificationClient;
+    private final UserService userService;
+    private final PriorityQueue<QueueItem> queue = new PriorityQueue<>(
+            (a, b) -> a.getAddedAt().compareTo(b.getAddedAt())
+    );
+
+    public void addToQueue(Long cartItemId) {
+        removeFromQueue(cartItemId);
+
+        QueueItem queueItem = new QueueItem(LocalDateTime.now(), cartItemId);
+        queue.offer(queueItem);
+
+    }
+
+    public void processQueue() {
+        LocalDateTime now = LocalDateTime.now();
+
+        while (!queue.isEmpty() && queue.peek().getAddedAt().plusMinutes(1).isBefore(now)) {
```

Comment:
make `CartItemReservationConfiguration` class which will have @ Configuration (spring config) annotation and one of the configuration is `reservationTimeInMinutes`

```
while (queueHasExpiredItemReservations()) {}

...
...

private boolean queueHasExpiredItemReservations() {
      return !queue.isEmpty() && queue.peek().getAddedAt().plusMinutes(reservationTimeInMinutes).isBefore(now);
}
```